servant/src/Servant/API/QueryParam.hs

```diff
@@ -38,6 +38,37 @@ data QueryParam' (mods :: [*]) (sym :: Symbol) (a :: *)
 data QueryParams (sym :: Symbol) (a :: *)
     deriving Typeable
 
+-- | Lookup the values associated to the a collection of query string parameters
+-- and try to extract them as a value of type @a@. This is typically
+-- meant to query string parameters of the form
+-- @param1=val1&param2=val2@ and so on into a custom type represented by the form.
+--
+-- Note: Unlike with 'QueryParam', by default 'QueryParamForm' is parsed in a
+-- 'Lenient' way, because it's difficult to know if it should be parsed
+-- or not (when other 'QueryParam's are present). As a result, most users
+-- of 'QueryParamForm' are going to implement handlers that take a value
+-- of type (Maybe (Either Text a))
+-- 
+-- Example:
+--
+-- > data BookSearchParams = BookSearchParams
+-- >   { title :: Text
+-- >   , authors :: [Text]
+-- >   , page :: Maybe Int
+-- >   } deriving (Eq, Show, Generic)
+-- > instance FromForm BookSearchParams
+-- > type MyApi = "books" :> QueryParamForm "searchQ" BookSearchParams :> Get '[JSON] [Book]
+-- 
+-- Example Handler Signature:
+-- Maybe (Either Text BookSearchParams) -> Handler [Book]
+
+type QueryParamForm = QueryParamForm' '[Optional, Lenient]
```

Comment:
Sorry that I'm late to the party. Would everyhting be simpler if only `Lenient, Strict` is considered for `QueryParamForm'`. If i understand correctly, we **always** try to parse the form, so there is no `Optional / Required` dimension?

Reply:
I was wondering about something like that: would it ever make sense to have `'[Required, Lenient]` or `'[Required, Strict]` for this modifier?

When you ask if it would be simpler, do you mean from the perspective of a `Handler` which would presumably receive `Either a` instead of `Maybe (Either a)`? I did consider this from the perspective of a new Servant user: it *might* be confusing. On the other hand, it's more Servant-style (imho) to give users freedom and flexibility: if they really want `Required` or `Strict`, it wouldn't be tough to make an alias like the above.

I would defer to you guys on these and other questions. My gut says that going with @alpmestan's thinking and giving users freedom at the cost of a slightly more layered argument `Maybe (Either a)` is the right way to go.

Reply:
Yes. I am still not really satisfied with that implementation (I mentioned this in my first comment, *way up there*). The only alternative that occurred to me was insisting on a `Data.Data` instance and then looking for some keys from the record in the query string. I can add more text to the documentation. Let me know if you think more is needed.

Reply:
`Data.Data` won't work. There are no guarantee that the data type resembles query parameters (simple renamings ruins that).

Note: there might be `FromForm` which succeeds on empty `Form` (e.g. all fields are optional). It's a small gotcha, which we should document. It shouldn't be a big issue.